services/motion/builtin/move_request.go

```diff
@@ -85,12 +90,88 @@ func (mr *moveRequest) deviatedFromPlan(ctx context.Context, waypoints [][]refer
 	if err != nil {
 		return false, err
 	}
-	mr.planRequest.Logger.Debug("deviation from plan: %v", errorState.Point())
+	mr.planRequest.Logger.Debugf("deviation from plan: %v", errorState.Point())
 	return errorState.Point().Norm() > mr.config.PlanDeviationMM, nil
 }
 
 func (mr *moveRequest) obstaclesIntersectPlan(ctx context.Context, waypoints [][]referenceframe.Input, waypointIndex int) (bool, error) {
-	// TODO(RSDK-4507): implement this function
+	// convert waypoints to type of motionplan.Plan
+	var plan []map[string][]referenceframe.Input
+	// We only care to check against waypoints we have not reached yet.
+	for _, inputs := range waypoints[waypointIndex:] {
+		input := make(map[string][]referenceframe.Input)
+		input[mr.kinematicBase.Name().Name] = inputs
+		plan = append(plan, input)
+	}
+
+	// iterate through vision services
+	for _, srvc := range mr.visionSvcs {
+		// iterate through all cameras on the robot
+		// any camera may be used by any vision service
+		for _, camName := range mr.cameraNames {
+			// get detections from vision service
+			detections, err := srvc.GetObjectPointClouds(ctx, camName.Name, nil)
+			if err != nil {
+				return false, err
```

Comment:
Isn't this highly likely to error if multiple cameras and services are used?

Reply:
I may have a different understanding of this question than you..
But yes, while any vision service my use "any" camera available to it, it is also the case that not all vision services actually implement `GetObjectPointClouds`.
If this is the case, then our returned error will be will be of form
```
errors.Errorf("vision model %q does not implement a 3D segmenter", vm.Named.Name().String())
```
I will add a check for this so that we do not error out entirely.

Reply:
What does returning an error here do? Will it trigger replanning? I think it should based on the scope ("If motion can’t get responses from `GetObjectPointCloud` at N Hz, motion should log a descriptive error, and stop the current plan and create a new plan.").

Reply:
In both cases since we return an error which is not `context.Cancelled`, the `MoveOnGlobe` call fails and errors out.

In the case where we error because the vision model itself does not implement a 3D segmenter, I do not think we should stop and replan since we would always experience the same error.
If we have only one obstacle detector and we experience this error, I think it makes more sense to stop the execution of the call entirely since we would not even be able to perform obstacle detection.

In the case where the error is something else, similarly, I am not sure if it makes sense to stop the execution of the current plan in favor of starting/creating a new one since the same error would persist.

what do you think?

Reply:
I agree that if the vision model does not implement a 3D segmenter, then it doesn't make sense to stop and replan.

For an unknown error, we should do what's in the scope: if we can't access obstacles, then we stop and replan. This might mean we get stuck replanning in a loop, which is fine. Or the error might resolve, and replanning will succeed.

Reply:
kk, I will fix this.

Reply:
I'm not sure I like simply skipping over unimplemented errors, even with a logger message.

I would propose that at initialization of planning, we should call `GetObjectPointCloud` on every camera-vision pair, and if there's an error then that should be reported and stop the whole thing, so a user doesn't think that they might have a working vision service (especially when not in debug mode) when they do not because they specified an invalid viz service. Especially because they could be using any vision service, it doesn't need to return the string "does not implement a 3D segmenter" in the error.

@tessavitabile if we do that, at what point to we actually error everything rather than replanning? E.g. if a camera gets unplugged, then the viz service will error continuously, presumably we would want to actually fail the move call at some point rather than continuously trying to plan and being interrupted by the camera error.

Reply:
Actually looks like my question was answered while I was typing

> This might mean we get stuck replanning in a loop, which is fine. Or the error might resolve, and replanning will succeed.

Reply:
I think the reason why we scoped that an error should cause replanning is because of the navigation service behavior of moving on to the next waypoint if it encounters an error. Replanning in a loop seems preferable to having the navigation service ignore the failure and continue to the next waypoint.

In the future, we'll probably want to add support to the navigation service for distinguishing between different kinds of errors and deciding what to do (retry the current waypoint, move on to the next waypoint, enter some failure state).